accepted/future-releases/class-modifiers/feature-specification.md

```diff
@@ -836,38 +785,82 @@ the latter, even if the class is being used as a mixin in a post-feature
+anonymous mixin application classes elsewhere, we infer class modifiers as
+follows.
+
+First we define the following transitive property on declarations:
+
+* A declaration *prohibits inheritance* _(of its implementation)_ if:
+
+  * the declaration is marked `interface` or `final`, or
+  * the declaration is marked `sealed` and extends or mixes in
+    a declaration which prohibits inheritance.
+
+Let then *C* be an anonymous mixin application occurring in
+a post-feature library, with superclass *S* and mixin *M*.
 
-To ensure reasonable and correct behavior, we infer class modifiers on anonymous
-mixin application classes as follows.
+* If either *S* or *M* has a `sealed` modifier, then *C* has an implicit
+  `sealed` modifier.
 
-Let *C* be an anonymous mixin application with superclass *S* and mixin *M*. Then:
+* Otherwise *C* is `abstract`, and
 
-* If any of *S* or *M* has a `sealed` modifier, *C* is has a `sealed` modifier.
-* Otherwise:
-  * *C* is `abstract`, and
-  * If either of *S* or *M* has a `base` or `final` modifier, then *C* has a `final` modifier.
+  * If *C* *prohibits inheritance* and its interface *prevents implementation*,
+    then *C* has an implicit `final` modifier.
 
-_We do not distinguish whether *S* or *M* has `base` or `final` modifiers.
-The modifier on *C* is there to satisfy the requirement that a subtype of a `base` or `final`
-declaration is itself `base`, `final` or `sealed`. The anonymous mixin application class will
-be immediately extended inside the same library, which is allowed by both `base` and `final`.,
-and will not be used for anything else._
+  * Otherwise if *C* *prevents implementation*, then *C* has an implicit
+    `base` modifier.
 
-Adding `sealed` to an anonymous mixin application class with only one subclass ensures
-that the subclass extending the mixin application class can be used
-in exhaustiveness checking of the sealed superclass.
-This is necessary since the anonymous mixin application class itself cannot be referenced.
+  * Otherwise if *C* *prohibits inheritance*, then *C* has an implicit
+   `interface` modifier.
+
+  * Otherwise *C* has no modifier.
+
+Adding `sealed` to an anonymous mixin application class, which always has
+precisely one subclass, ensures that the subclass can be used in
+exhaustiveness checking of the sealed superclass.
+
+Adding `final` or `base` satisfies the requirement that a subtype of a
+`base` or `final` declaration is itself `base`, `final` or `sealed`.
+
+Adding `interface` ensures that the mixin application class doesn’t
+remove a restriction on the implementation inherited from *S* or *M*.
+This makes the “reopen” lint described below easier to implement.
+
+Adding these modifiers on the anonymous mixin application class ensures that
+the anonymous class can mostly be ignored, since it satisfies all possible
+requirements of its superclasses, while still propagating those requirements
+to its subclass.
```

Comment:
That's a whole lot of specification just to reach the point where anonymous mixin applications can now disappear back into the shadows where they belong. :)

Reply:
Yep. But maybe necessary.

Reply:
Narrator voice: They weren't.

(Huzzah!)